proposals/type-classes.md

```diff
@@ -267,12 +275,24 @@ Call site:
 fun addInts(): Int = add(1, 2)
 ```
 
-1. The compiler first looks at the file the interface is defined in. If it finds exactly one implementation it uses that instance.
-2. If it fails to find an implementation in the interface's file, it then looks at the files of the implemented types. For each type class parameter check the file it was defined in. If exactly one implementation is found use that instance.
+1. The compiler first looks at the function context where the invocation is happening. If a function argument matches the required instance for a typeclass, it uses that instance; e.g.:
+
+```kotlin
+fun <a> duplicate(a : A, with M: Monoid<A>): A = a.combine(a)
+```
+
+The invocation `a.combine(a)` requires a `Monoid<A>` and since one is passed as an argument to `duplicate`, it uses that one.
+
+2. In case it fails, it inspects the following places, sequentially, until it is able to find a valid unique instance for the typeclass:
+    a. The current package (where the invocation is taking place)
```

Comment:
if we allow instance at invocation site we are allowing orphan instance and based on the original keep those should not be supported. If anything these would have to be flagged as `internal` in which case they'd have precedence over canonical ones and would be non-exportable. If that is not the case is looking in the invocation package necessary?

Reply:
It is not necessary and it is open to discussion. The current implementation is able to search for an instance in that place, but if that causes troubles with orphan instances, we can just remove it.

Reply:
We should either remove it or leave it allowing orphan instances with `internal` that are not exportable

Reply:
Good. I'll check that and add it for the next version.